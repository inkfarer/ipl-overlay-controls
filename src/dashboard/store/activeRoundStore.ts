import { ActiveRound, SwapColorsInternally } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { GameWinner } from 'types/enums/gameWinner';
import { SetActiveColorRequest, SetWinnerRequest, SwapRoundColorRequest } from 'types/messages/activeRound';
import { SetRoundRequest } from 'types/messages/rounds';
import { ActiveRoundGame } from 'types/activeRoundGame';

const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export const activeRoundReps = [ activeRound, swapColorsInternally ];

export interface ActiveRoundStore {
    activeRound: ActiveRound;
    swapColorsInternally: SwapColorsInternally;
}

export const activeRoundStore = createStore<ActiveRoundStore>({
    state: {
        activeRound: null,
        swapColorsInternally: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        setWinner(store, { winner }: { winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner });
        },
        removeWinner(): void {
            nodecg.sendMessage('removeWinner');
        },
        setActiveColor(store, request: SetActiveColorRequest): void {
            nodecg.sendMessage('setActiveColor', request);
        },
        swapColors(): void {
            swapColorsInternally.value = !swapColorsInternally.value;
        },
        setWinnerForIndex(store, { index, winner }: { index: number, winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner, roundIndex: index } as SetWinnerRequest);
        },
        resetActiveRound(): void {
            nodecg.sendMessage('resetActiveRound');
        },
        setActiveRound(store, request: SetRoundRequest): void {
            nodecg.sendMessage('setActiveRound', request);
        },
        swapRoundColor(store, request: SwapRoundColorRequest): void {
            nodecg.sendMessage('swapRoundColor', request);
        },
        updateActiveGames(store, games: ActiveRoundGame[]): void {
            nodecg.sendMessage('updateActiveGames', { games });
        },
        switchToNextColor(): void {
            nodecg.sendMessage('switchToNextColor');
        },
        switchToPreviousColor(): void {
            nodecg.sendMessage('switchToPreviousColor');
        },
        async getNextAndPreviousColors(): Promise<void> {
            return nodecg.sendMessage('getNextAndPreviousColors');
        }
    }
});

export const activeRoundStoreKey: InjectionKey<Store<ActiveRoundStore>> = Symbol();

export function useActiveRoundStore(): Store<ActiveRoundStore> {
    return useStore(activeRoundStoreKey);
}
