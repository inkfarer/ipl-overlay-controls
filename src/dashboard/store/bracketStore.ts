import { Bracket } from '@tourneyview/common';
import { defineStore } from 'pinia';

const bracketData = nodecg.Replicant<Bracket | undefined>('bracketData');

export const bracketReps = [ bracketData ];

export interface BracketStore {
    bracketData: Bracket | undefined;
}

export const useBracketStore = defineStore('brackets', {
    state: () => ({
        bracketData: undefined
    } as BracketStore),
    actions: {
        resetBracketData() {
            bracketData.value = undefined;
        }
    }
});
