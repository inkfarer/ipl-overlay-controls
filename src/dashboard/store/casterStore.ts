import { NodeCGBrowser } from 'nodecg/browser';
import { Caster, Casters, RadiaSettings } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { generateId } from '../../helpers/generateId';

const casters = nodecg.Replicant<Casters>('casters');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');

export const castersReps = [ casters, radiaSettings ];

export interface CasterStore {
    casters: Casters;
    uncommittedCasters: Casters;
    radiaSettings: RadiaSettings;
}

const defaultCaster = {
    name: 'New Caster',
    twitter: '',
    pronouns: '?/?'
};

export const casterStore = createStore<CasterStore>({
    state: {
        casters: {},
        uncommittedCasters: {},
        radiaSettings: {
            enabled: null,
            guildID: null,
            updateOnImport: null
        }
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        },
        updateCaster(store, { id, newValue }: { id: string, newValue: Caster }): void {
            casters.value[id] = newValue;
        },
        addUncommittedCaster(store, { id, caster }: { id: string, caster?: Caster }): void {
            store.uncommittedCasters[id] = caster ?? defaultCaster;
        },
        removeUncommittedCaster(store, id: string): void {
            delete store.uncommittedCasters[id];
        }
    },
    actions: {
        addUncommittedCaster(store): string {
            const id = generateId();
            store.commit('addUncommittedCaster', { id });
            return id;
        },
        async saveUncommittedCaster(store, { id, caster }: { id: string, caster: Caster }): Promise<string> {
            const newId = await nodecg.sendMessage('saveCaster', caster);
            store.commit('removeUncommittedCaster', id);
            return newId;
        },
        async loadCastersFromVc(store): Promise<void> {
            const result = await nodecg.sendMessage('getLiveCommentators');
            if (result.extra && result.extra.length > 0) {
                for (let i = 0; i < result.extra.length; i++) {
                    const extraCaster = result.extra[i];
                    const id = extraCaster.discord_user_id;
                    delete extraCaster.discord_user_id;
                    store.commit('addUncommittedCaster', { id, caster: extraCaster });
                }
            }
        },
        showCasters() {
            nodecg.sendMessage('mainShowCasters');
        },
        removeCaster(store, id: string): void {
            nodecg.sendMessage('removeCaster', { id });
        }
    }
});

export const casterStoreKey: InjectionKey<Store<CasterStore>> = Symbol();

export function useCasterStore(): Store<CasterStore> {
    return useStore(casterStoreKey);
}
