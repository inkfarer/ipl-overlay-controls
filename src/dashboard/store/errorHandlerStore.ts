import { createStore, Store, useStore } from 'vuex';
import { App, InjectionKey } from 'vue';
import { generateId } from '../../helpers/generateId';

export interface ErrorHandlerStore {
    recentErrors: Record<string, unknown>;
}

export const errorHandlerStore = createStore<ErrorHandlerStore>({
    state: {
        recentErrors: {}
    },
    mutations: {
        removeRecentError(state, { key }: { key: string }): void {
            delete state.recentErrors[key];
        }
    },
    actions: {
        handleError(store, { err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err);

            if (Object.keys(store.state.recentErrors).length >= 2) return;

            const id = generateId();
            store.state.recentErrors[id] = err;
            window.setTimeout(() => {
                delete store.state.recentErrors[id];
            }, 25000);
        }
    }
});

export const errorHandlerStoreKey: InjectionKey<Store<ErrorHandlerStore>> = Symbol();

export function useErrorHandlerStore(): Store<ErrorHandlerStore> {
    return useStore(errorHandlerStoreKey);
}

export function setUpErrorHandler(app: App<unknown>): void {
    app.use(errorHandlerStore, errorHandlerStoreKey);
    app.config.errorHandler = (err, vm, info) => {
        errorHandlerStore.dispatch('handleError', { err, info });
    };
}
