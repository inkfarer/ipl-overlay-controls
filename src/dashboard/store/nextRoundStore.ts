import { NextRound } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetRoundRequest } from 'types/messages/rounds';

const nextRound = nodecg.Replicant<NextRound>('nextRound');

export const nextRoundReps = [ nextRound ];

export interface NextRoundStore {
    nextRound: NextRound;
}

export const nextRoundStore = createStore<NextRoundStore>({
    state: {
        nextRound: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        },
        setShowOnStream(store, newValue: boolean) {
            nextRound.value.showOnStream = newValue;
        }
    },
    actions: {
        beginNextMatch(): void {
            nodecg.sendMessage('beginNextMatch');
        },
        setNextRound(store, {
            teamAId,
            teamBId,
            roundId
        }: { teamAId: string, teamBId: string, roundId: string }): void {
            nodecg.sendMessage('setNextRound', { teamAId, teamBId, roundId } as SetRoundRequest);
        },
    }
});

export const nextRoundStoreKey: InjectionKey<Store<NextRoundStore>> = Symbol();

export function useNextRoundStore(): Store<NextRoundStore> {
    return useStore(nextRoundStoreKey);
}
