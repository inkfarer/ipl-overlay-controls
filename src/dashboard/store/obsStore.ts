import { NodeCGBrowser } from 'nodecg/browser';
import { ObsCredentials, ObsData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetObsDataRequest } from 'types/messages/obs';

const obsData = nodecg.Replicant<ObsData>('obsData');
const obsCredentials = nodecg.Replicant<ObsCredentials>('obsCredentials');

export const obsReps = [obsData, obsCredentials];

export interface ObsStore {
    obsData: ObsData
    obsCredentials: ObsCredentials
}

export const obsStore = createStore<ObsStore>({
    state: {
        obsData: null,
        obsCredentials: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async connect(store, { address, password }: { address: string, password?: string }): Promise<void> {
            return nodecg.sendMessage('connectToObs', { address, password });
        },
        async setData(store, data: SetObsDataRequest): Promise<void> {
            return nodecg.sendMessage('setObsData', data);
        },
        async startGame(): Promise<void> {
            return nodecg.sendMessage('startGame');
        },
        async endGame(): Promise<void> {
            return nodecg.sendMessage('endGame');
        },
        setEnabled(store, enabled: boolean): void {
            nodecg.sendMessage('setObsSocketEnabled', enabled);
        }
    }
});

export const obsStoreKey: InjectionKey<Store<ObsStore>> = Symbol();

export function useObsStore(): Store<ObsStore> {
    return useStore(obsStoreKey);
}
