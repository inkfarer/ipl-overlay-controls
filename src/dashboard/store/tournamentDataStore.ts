import { RoundStore, TournamentData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { ToggleTeamImageRequest } from 'types/messages/tournamentData';
import { UpdateRoundStoreRequest, UpdateRoundStoreResponse } from 'types/messages/roundStore';

const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');
const roundStore = nodecg.Replicant<RoundStore>('roundStore');

export const tournamentDataReps = [ tournamentData, roundStore ];

export interface TournamentDataStore {
    tournamentData: TournamentData;
    roundStore: RoundStore;
}

export const tournamentDataStore = createStore<TournamentDataStore>({
    state: {
        tournamentData: null,
        roundStore: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async getTournamentData(store, { method, id }: { method: TournamentDataStore, id: string }): Promise<void> {
            return await nodecg.sendMessage('getTournamentData', { method, id });
        },
        async getSmashggEvent(store, { eventId }: { eventId: number }): Promise<void> {
            return await nodecg.sendMessage('getSmashggEvent', { eventId });
        },
        async uploadTeamData(store, { file }: { file: File }): Promise<void> {
            return sendLocalFile('teams', file);
        },
        async uploadRoundData(store, { file }: { file: File }): Promise<void> {
            return sendLocalFile('rounds', file);
        },
        async fetchRoundData(store, { url }: { url: string }): Promise<void> {
            return nodecg.sendMessage('getRounds', { url });
        },
        async updateRound(store, data: UpdateRoundStoreRequest): Promise<UpdateRoundStoreResponse> {
            return nodecg.sendMessage('updateRoundStore', data);
        },
        resetRoundStore() {
            nodecg.sendMessage('resetRoundStore');
        },
        setTeamImageHidden(store, { teamId, isVisible }: { teamId: string, isVisible: boolean }): Promise<void> {
            return nodecg.sendMessage('toggleTeamImage', { teamId, isVisible } as ToggleTeamImageRequest);
        },
        removeRound(store, { roundId }: { roundId: string }): Promise<void> {
            return nodecg.sendMessage('removeRound', { roundId });
        }
    }
});

async function sendLocalFile(dataType: 'teams' | 'rounds', file: File): Promise<void> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('jsonType', dataType);

    const response = await fetch('/ipl-overlay-controls/upload-tournament-json', { method: 'POST', body: formData });

    if (response.status !== 200) {
        const responseText = await response.text();
        throw new Error(`Import failed with status ${response.status}: ${responseText}`);
    }
}

export const tournamentDataStoreKey: InjectionKey<Store<TournamentDataStore>> = Symbol();

export function useTournamentDataStore(): Store<TournamentDataStore> {
    return useStore(tournamentDataStoreKey);
}
