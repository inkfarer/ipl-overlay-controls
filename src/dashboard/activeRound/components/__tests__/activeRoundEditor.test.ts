import ActiveRoundEditor from '../activeRoundEditor.vue';
import { createStore } from 'vuex';
import { ActiveRoundStore, activeRoundStoreKey } from '../../../store/activeRoundStore';
import { GameWinner } from 'types/enums/gameWinner';
import { TournamentDataStore, tournamentDataStoreKey } from '../../../store/tournamentDataStore';
import { config, mount } from '@vue/test-utils';

describe('ActiveRoundEditor', () => {
    config.global.stubs = {
        IplSelect: true,
        IplCheckbox: true,
        FontAwesomeIcon: true,
        IplButton: true
    };

    const mockSetActiveColor = jest.fn();
    const mockSetActiveRound = jest.fn();
    const mockSwapColors = jest.fn();

    function createActiveRoundStore() {
        return createStore<ActiveRoundStore>({
            state: {
                activeRound: {
                    teamA: {
                        score: 0,
                        id: '123123',
                        name: null,
                        showLogo: true,
                        players: null,
                        color: null
                    },
                    teamB: {
                        score: 2,
                        id: '345345',
                        name: null,
                        showLogo: false,
                        players: null,
                        color: null
                    },
                    activeColor: {
                        categoryName: 'Ranked Modes',
                        index: 0,
                        title: 'coolest color',
                        isCustom: false
                    },
                    round: {
                        id: '0387',
                        name: 'cool round',
                        isCompleted: false
                    },
                    games: [
                        {
                            winner: GameWinner.BRAVO,
                            stage: null,
                            mode: null
                        },
                        {
                            winner: GameWinner.BRAVO,
                            stage: null,
                            mode: null
                        },
                        {
                            winner: GameWinner.NO_WINNER,
                            stage: null,
                            mode: null
                        },
                    ],
                },
                swapColorsInternally: false
            },
            actions: {
                swapColors: mockSwapColors,
                setActiveColor: mockSetActiveColor,
                setActiveRound: mockSetActiveRound
            }
        });
    }

    function createTournamentDataStore() {
        return createStore<TournamentDataStore>({
            state: {
                tournamentData: {
                    meta: { id: '1093478', source: 'SMASHGG' },
                    teams: [
                        { id: '123123', name: 'cool team A (test long name long name long name long name long name long name long name)', players: [], showLogo: true },
                        { id: '345345', name: 'cool team B', players: [], showLogo: false }
                    ]
                },
                roundStore: {
                    '0387': {
                        meta: { name: 'dope round', isCompleted: false },
                        games: []
                    }
                }
            },
            actions: {
                setTeamImageHidden: jest.fn()
            }
        });
    }

    it('matches snapshot and has expected values for select options', () => {
        const tournamentDataStore = createTournamentDataStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
        const teamASelector = wrapper.getComponent('[data-test="team-a-selector"]');
        expect(teamASelector.attributes().modelvalue).toEqual('123123');
        expect((teamASelector.vm.$props as { options: unknown }).options).toEqual([
            { name: 'cool team A (test long name long name long na...', value: '123123' },
            { name: 'cool team B', value: '345345' }
        ]);
        const teamBSelector = wrapper.getComponent('[data-test="team-b-selector"]');
        expect(teamBSelector.attributes().modelvalue).toEqual('345345');
        expect((teamBSelector.vm.$props as { options: unknown }).options).toEqual([
            { name: 'cool team A (test long name long name long na...', value: '123123' },
            { name: 'cool team B', value: '345345' }
        ]);
        const roundSelector = wrapper.getComponent('[data-test="round-selector"]');
        expect(roundSelector.attributes().modelvalue).toEqual('0387');
        expect((roundSelector.vm.$props as { options: unknown }).options).toEqual([
            { name: 'dope round', value: '0387' }
        ]);
    });

    it('handles store data updating', async () => {
        const tournamentDataStore = createTournamentDataStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });

        activeRoundStore.state.activeRound.teamA.id = '678678';
        await wrapper.vm.$nextTick();

        expect(wrapper.getComponent('[data-test="team-a-selector"]').attributes().modelvalue).toEqual('678678');
    });

    it('has expected update button color if data is locally changed', async () => {
        const tournamentDataStore = createTournamentDataStore();
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });
        const teamASelector = wrapper.getComponent('[data-test="team-a-selector"]');

        teamASelector.vm.$emit('update:modelValue', '098098');
        await wrapper.vm.$nextTick();

        expect(wrapper.getComponent('[data-test="update-round-button"]').attributes().color).toEqual('red');
    });

    it('sends commit to store if team A image toggle is changed', () => {
        const tournamentDataStore = createTournamentDataStore();
        jest.spyOn(tournamentDataStore, 'dispatch');
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });
        const teamAImageToggle = wrapper.getComponent('[data-test="team-a-image-toggle"]');

        teamAImageToggle.vm.$emit('update:modelValue', false);

        expect(tournamentDataStore.dispatch).toHaveBeenCalledWith('setTeamImageHidden', {
            teamId: '123123',
            isVisible: false
        });
    });

    it('sends commit to store if team B image toggle is changed', () => {
        const tournamentDataStore = createTournamentDataStore();
        jest.spyOn(tournamentDataStore, 'dispatch');
        const activeRoundStore = createActiveRoundStore();
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });
        const teamBImageToggle = wrapper.getComponent('[data-test="team-b-image-toggle"]');

        teamBImageToggle.vm.$emit('update:modelValue', true);

        expect(tournamentDataStore.dispatch).toHaveBeenCalledWith('setTeamImageHidden', {
            teamId: '345345',
            isVisible: true
        });
    });

    it('updates round data on update button click', () => {
        const tournamentDataStore = createTournamentDataStore();
        const activeRoundStore = createActiveRoundStore();
        jest.spyOn(activeRoundStore, 'dispatch');
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });
        const updateButton = wrapper.getComponent('[data-test="update-round-button"]');

        updateButton.vm.$emit('click');

        expect(activeRoundStore.dispatch).toHaveBeenCalledWith('setActiveRound', {
            teamAId: '123123',
            teamBId: '345345',
            roundId: '0387'
        });
    });

    it('does not display message if selected round has no progress and round data is unchanged', () => {
        const activeRoundStore = createActiveRoundStore();
        const tournamentDataStore = createTournamentDataStore();
        tournamentDataStore.state.roundStore = {
            '0387': {
                meta: { name: 'dope round', isCompleted: false },
                teamA: { id: '123123', name: 'Cool Team', score: 0, showLogo: true, players: []},
                teamB: { id: '345345', name: 'Cool Team 2', score: 0, showLogo: true, players: []},
                games: []
            }
        };
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });

        expect(wrapper.findComponent('[data-test="round-progress-message"]').exists()).toEqual(false);
    });

    it('does not display message if selected round has no progress', async () => {
        const activeRoundStore = createActiveRoundStore();
        const tournamentDataStore = createTournamentDataStore();
        tournamentDataStore.state.roundStore = {
            '0387': {
                meta: { name: 'dope round', isCompleted: false },
                teamA: { id: '123123', name: 'Cool Team', score: 0, showLogo: true, players: []},
                teamB: { id: '345345', name: 'Cool Team 2', score: 0, showLogo: true, players: []},
                games: []
            }
        };
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });
        wrapper.getComponent('[data-test="team-a-selector"]').vm.$emit('update:modelValue', '098098');
        await wrapper.vm.$nextTick();

        expect(wrapper.findComponent('[data-test="round-progress-message"]').exists()).toEqual(false);
    });

    it('does not display message if selected round has progress but data is not changed', () => {
        const activeRoundStore = createActiveRoundStore();
        const tournamentDataStore = createTournamentDataStore();
        tournamentDataStore.state.roundStore = {
            '0387': {
                meta: { name: 'dope round', isCompleted: false },
                teamA: { id: '123123', name: 'Cool Team', score: 0, showLogo: true, players: []},
                teamB: { id: '345345', name: 'Cool Team 2', score: 1, showLogo: true, players: []},
                games: []
            }
        };
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });

        expect(wrapper.findComponent('[data-test="round-progress-message"]').exists()).toEqual(false);
    });

    it('displays message if selected round has progress', async () => {
        const activeRoundStore = createActiveRoundStore();
        const tournamentDataStore = createTournamentDataStore();
        tournamentDataStore.state.roundStore = {
            '0387': {
                meta: { name: 'dope round', isCompleted: false },
                teamA: { id: '123123', name: 'Cool Team (long name long name long name long name long name)', score: 0, showLogo: true, players: []},
                teamB: { id: '345345', name: 'Cool Team 2', score: 1, showLogo: true, players: []},
                games: []
            }
        };
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });

        wrapper.getComponent('[data-test="team-a-selector"]').vm.$emit('update:modelValue', '098098');
        await wrapper.vm.$nextTick();

        const roundProgressMessage = wrapper.findComponent('[data-test="round-progress-message"]');
        expect(roundProgressMessage.exists()).toEqual(true);
        expect(roundProgressMessage.text()).toEqual('\'dope round\' already has saved progress. (Cool Team (long name long name long name long... vs Cool Team 2)');
    });

    it('displays message if selected round is completed', async () => {
        const activeRoundStore = createActiveRoundStore();
        const tournamentDataStore = createTournamentDataStore();
        tournamentDataStore.state.roundStore = {
            '0387': {
                meta: { name: 'dope round', isCompleted: true },
                teamA: { id: '123123', name: 'Cool Team', score: 2, showLogo: true, players: []},
                teamB: { id: '345345', name: 'Cool Team 2', score: 1, showLogo: true, players: []},
                games: []
            }
        };
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });

        wrapper.getComponent('[data-test="team-a-selector"]').vm.$emit('update:modelValue', '098098');
        await wrapper.vm.$nextTick();

        const roundProgressMessage = wrapper.findComponent('[data-test="round-progress-message"]');
        expect(roundProgressMessage.exists()).toEqual(true);
        expect(roundProgressMessage.text()).toEqual('\'dope round\' is already completed. (Cool Team vs Cool Team 2)');
    });

    it('changes selected teams if a new round with progress is selected', async () => {
        const activeRoundStore = createActiveRoundStore();
        const tournamentDataStore = createTournamentDataStore();
        tournamentDataStore.state.roundStore = {
            '0387': {
                meta: { name: 'dope round', isCompleted: true },
                teamA: { id: '123123', name: 'Cool Team', score: 2, showLogo: true, players: []},
                teamB: { id: '345345', name: 'Cool Team 2', score: 1, showLogo: true, players: []},
                games: []
            },
            '12345': {
                meta: { name: 'dope round 2', isCompleted: false },
                teamA: { id: '789789', name: 'Cool Team 3', score: 2, showLogo: true, players: []},
                teamB: { id: '678678', name: 'Cool Team 4', score: 1, showLogo: true, players: []},
                games: []
            }
        };
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });
        const roundSelector = wrapper.getComponent('[data-test="round-selector"]');

        roundSelector.vm.$emit('update:modelValue', '12345');
        await wrapper.vm.$nextTick();

        expect(wrapper.findComponent('[data-test="round-progress-message"]').text()).toEqual('\'dope round 2\' already has saved progress. (Cool Team 3 vs Cool Team 4)');
        expect(wrapper.getComponent('[data-test="team-a-selector"]').attributes().modelvalue).toEqual('789789');
        expect(wrapper.getComponent('[data-test="team-b-selector"]').attributes().modelvalue).toEqual('678678');
    });

    it('changes selected teams to active round teams if a new round without progress is selected', async () => {
        const activeRoundStore = createActiveRoundStore();
        activeRoundStore.state.activeRound = {
            teamA: { id: '567234', name: 'cool team G', showLogo: true, players: [], score: 0, color: null },
            teamB: { id: '123098', name: 'cool team Z', showLogo: false, players: [], score: 0, color: null },
            round: { id: '0387', name: 'dope round', isCompleted: false },
            games: [],
            activeColor: {
                categoryName: 'Ranked Modes',
                index: 0,
                title: 'coolest color',
                isCustom: false
            },
        };
        const tournamentDataStore = createTournamentDataStore();
        tournamentDataStore.state.roundStore = {
            '0387': {
                meta: { name: 'dope round', isCompleted: true },
                teamA: { id: '123123', name: 'Cool Team', score: 2, showLogo: true, players: []},
                teamB: { id: '345345', name: 'Cool Team 2', score: 1, showLogo: true, players: []},
                games: []
            },
            '12345': {
                meta: { name: 'dope round 2', isCompleted: false },
                teamA: { id: '789789', name: 'Cool Team 3', score: 2, showLogo: true, players: []},
                teamB: { id: '678678', name: 'Cool Team 4', score: 1, showLogo: true, players: []},
                games: []
            },
            '123456': {
                meta: { name: 'dope round 3', isCompleted: false },
                games: []
            }
        };
        const wrapper = mount(ActiveRoundEditor, {
            global: {
                plugins: [[tournamentDataStore, tournamentDataStoreKey], [activeRoundStore, activeRoundStoreKey]]
            }
        });
        const roundSelector = wrapper.getComponent('[data-test="round-selector"]');

        roundSelector.vm.$emit('update:modelValue', '12345');
        await wrapper.vm.$nextTick();
        roundSelector.vm.$emit('update:modelValue', '123456');
        await wrapper.vm.$nextTick();

        expect(wrapper.findComponent('[data-test="round-progress-message"]').exists()).toEqual(false);
        expect(wrapper.getComponent('[data-test="team-a-selector"]').attributes().modelvalue).toEqual('567234');
        expect(wrapper.getComponent('[data-test="team-b-selector"]').attributes().modelvalue).toEqual('123098');
    });
});
