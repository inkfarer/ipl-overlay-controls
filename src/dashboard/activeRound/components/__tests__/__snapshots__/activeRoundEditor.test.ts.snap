// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveRoundEditor matches snapshot and has expected values for select options 1`] = `
"<div class=\\"ipl-expansion-panel__content\\" data-test=\\"active-round-editor\\" data-v-40cedb04=\\"\\">
  <div class=\\"ipl-expansion-panel__header layout horizontal center-vertical\\" data-v-40cedb04=\\"\\">
    <div class=\\"ipl-expansion-panel__header-background\\" data-v-40cedb04=\\"\\"></div>
    <div class=\\"ipl-expansion-panel__title\\" data-v-40cedb04=\\"\\">Teams &amp; Rounds</div>
    <font-awesome-icon-stub icon=\\"chevron-left\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"icon\\" data-v-40cedb04=\\"\\"></font-awesome-icon-stub>
    <div class=\\"header-extra\\" data-v-40cedb04=\\"\\"></div>
  </div>
  <div class=\\"content\\" data-v-40cedb04=\\"\\" style=\\"display: none;\\">
    <!--v-if-->
    <div class=\\"layout horizontal\\">
      <div class=\\"layout vertical center-horizontal max-width\\">
        <ipl-select-stub modelvalue=\\"123123\\" label=\\"Team A\\" options=\\"[object Object],[object Object]\\" disabled=\\"false\\" data-test=\\"team-a-selector\\"></ipl-select-stub>
        <ipl-checkbox-stub label=\\"Show image\\" modelvalue=\\"true\\" small=\\"true\\" disabled=\\"false\\" class=\\"m-t-6\\" data-test=\\"team-a-image-toggle\\"></ipl-checkbox-stub>
      </div>
      <div class=\\"layout vertical center-horizontal max-width m-l-8\\">
        <ipl-select-stub modelvalue=\\"345345\\" label=\\"Team B\\" options=\\"[object Object],[object Object]\\" disabled=\\"false\\" data-test=\\"team-b-selector\\"></ipl-select-stub>
        <ipl-checkbox-stub label=\\"Show image\\" modelvalue=\\"false\\" small=\\"true\\" disabled=\\"false\\" class=\\"m-t-6\\" data-test=\\"team-b-image-toggle\\"></ipl-checkbox-stub>
      </div>
    </div>
    <ipl-select-stub modelvalue=\\"0387\\" label=\\"Round\\" options=\\"[object Object]\\" disabled=\\"false\\" class=\\"m-t-6\\" data-test=\\"round-selector\\"></ipl-select-stub>
    <ipl-button-stub label=\\"Update\\" color=\\"blue\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"false\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" class=\\"m-t-8\\" data-test=\\"update-round-button\\"></ipl-button-stub>
  </div>
</div>"
`;
